pkg/engine/plan/local_type_field_extractor.go

```diff
@@ -27,173 +27,263 @@ func NewLocalTypeFieldExtractor(document *ast.Document) *LocalTypeFieldExtractor
+		mutationType = "Mutation"
+	}
+	subscriptionType := string(e.document.Index.SubscriptionTypeName)
+	if subscriptionType == "" {
+		subscriptionType = "Subscription"
+	}
 
-	for i := range rootNodes {
-		fieldNameToRef := make(map[string]int, len(rootNodes[i].FieldNames))
+	// 1. Loop over each node in the document (see description above).
+	for _, astNode := range e.document.RootNodes {
+		var isInterface bool
+		var concreteTypeNames []string
+		typeName := e.document.NodeNameString(astNode)
 
-		rootNodeASTNode, exists := e.document.Index.FirstNodeByNameStr(rootNodes[i].TypeName)
-		if !exists {
+		switch astNode.Kind {
+		case ast.NodeKindObjectTypeDefinition, ast.NodeKindObjectTypeExtension:
+			for _, ref := range e.interfaceRefs(astNode) {
+				interfaceName := e.document.ResolveTypeNameString(ref)
+				// The document doesn't provide a way to directly look up the
+				// types that implement an interface, so instead we track the
+				// interfaces implemented for each type and after all nodes
+				// have been processed record the concrete types for each
+				// interface.
+				possibleInterfaceTypes[interfaceName] = append(
+					possibleInterfaceTypes[interfaceName], typeName)
+			}
+		case ast.NodeKindInterfaceTypeDefinition, ast.NodeKindInterfaceTypeExtension:
+			isInterface = true
+		case ast.NodeKindUnionTypeDefinition, ast.NodeKindUnionTypeExtension:
+			for _, ref := range e.unionMemberRefs(astNode) {
+				memberName := e.document.ResolveTypeNameString(ref)
+				concreteTypeNames = append(concreteTypeNames, memberName)
+			}
+		default:
 			continue
 		}
 
-		fieldRefs := e.document.NodeFieldDefinitions(rootNodeASTNode)
-		for _, fieldRef := range fieldRefs {
-			fieldName := e.document.FieldDefinitionNameString(fieldRef)
-			fieldNameToRef[fieldName] = fieldRef
+		nodeInfo, ok := nodeInfoMap[typeName]
+		if !ok {
+			nodeInfo = &nodeInformation{}
+			nodeInfoMap[typeName] = nodeInfo
 		}
 
-		for _, fieldName := range rootNodes[i].FieldNames {
-			fieldRef := fieldNameToRef[fieldName]
+		hasKey := e.NodeHasKeyDirective(astNode)
+		isFederationEntity := hasKey && !isInterface
 
-			fieldTypeName := e.document.NodeNameString(e.document.FieldDefinitionTypeNode(fieldRef))
-			e.findChildNodesForType(fieldTypeName, &childNodes)
-		}
-	}
+		isRootNode := typeName == queryType ||
+			typeName == mutationType ||
+			typeName == subscriptionType ||
+			isFederationEntity
 
-	return childNodes
-}
-
-func (e *LocalTypeFieldExtractor) findChildNodesForType(typeName string, childNodes *[]TypeField) {
-	node, exists := e.document.Index.FirstNodeByNameStr(typeName)
-	if !exists {
-		return
-	}
-
-	fieldsRefs := e.document.NodeFieldDefinitions(node)
+		nodeInfo.isInterface = isInterface
+		// A node may be a local extension of a root node. The node is
+		// considered a root node if ANY node related to the type is a root
+		// node.
+		nodeInfo.isRoot = nodeInfo.isRoot || isRootNode
+		nodeInfo.concreteTypeNames = concreteTypeNames
```

Comment:
I'm not really knowledgeable about any of the semantics here, but was excited to read this diff.  I was wondering: do you need to do a set-union here like you do on line 149?  I don't know what the semantics of unionMemberRefs is in the presence of union extensions.

Reply:
Yes, you're right. Here's the spec: https://spec.graphql.org/October2021/#sec-Union-Extensions. Union extensions DO NOT reiterate existing union members and can add new members. (This means a union is needed, but a set-union isn't necessary.) I'll update this code to collect the member types instead.

Reply:
There aren't many examples of union extensions. Here are a couple from the Apollo federation code base:
- https://github.com/apollographql/federation/blob/c7907dcecbb308565ca1ff370d70cd4ec1649c9d/internals-js/src/__tests__/definitions.test.ts#L494
- https://github.com/apollographql/federation/blob/c7907dcecbb308565ca1ff370d70cd4ec1649c9d/federation-js/src/composition/validate/postComposition/__tests__/providesFieldsSelectInvalidType.test.ts#L182